### PR TITLE
FOGL-2877: Initial implementation of DNP3 south plugin

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-2877: Initial implementation of DNP3 south plugin

Counter, Binary, BinaryOutputStatus and Analog objects are currettly supported.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+# -DFOGLAMP_INCLUDE
+# -DFOGLAMP_LIB
+# -DFOGLAMP_SRC
+# -DFOGLAMP_INSTALL
+#
+# If no -D options are given and FOGLAMP_ROOT environment variable is set
+# then FogLAMP libraries and header files are pulled from FOGLAMP_ROOT path.
+
+set(CMAKE_CXX_FLAGS "-std=c++11 -O3 -g")
+set(CMAKE_BUILD_TYPE "Debug")
```
Comment (MarkRiddoch) [inline CMakeLists.txt]: We should not checkin with Debug set here

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  The master link id for DNP3 protocol
+
+outstations
+  An array of DNP3 outstations.
+  
+  Configuration examples:
+
+.. code-block:: console
+
+    {"outstations":["IP_1:PORT_1:ID_1", "IP_2:PORT_2:ID_2"]}
```
Comment (MarkRiddoch) [inline README.rst]: Didn't you change this?

File: dnp3.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+	m_manager = manager;
+
+	this->unlockConfig();
+
+	// Set log levels
+	const auto logLevels = levels::NOTHING | flags::WARN | flags::ERR;
+
+	// Create TCP channel for outstation
+	std::shared_ptr<IChannel> channel =
+		manager->AddTCPClient("tcpclient_1", // alias in log messages
```
Comment (MarkRiddoch) [inline dnp3.cpp]: Could we make this more descriptive than tcpclient_1? Name use the service name.

File: dnp3.cpp [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+	Logger::getLogger()->info("configured DNP3 TCP outstation is: %s:%d, Link Id %d",
+				  outstation.address.c_str(),
+				  outstation.port,
+				  outstation.linkId);
+
+	// This object contains static configuration for the master, and transport/link layers
+	MasterStackConfig stackConfig;
+
+	// you can optionally override these defaults likei:
```
Comment (MarkRiddoch) [inline dnp3.cpp]: typo likei

File: dnp3.cpp [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+	// setting the application layer response timeout
+	// or change behaviors on the master
+	stackConfig.master.responseTimeout = TimeDuration::Seconds(5);
```
Comment (MarkRiddoch) [inline dnp3.cpp]: Maybe this should be a configuration option

File: dnp3.cpp [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+	if (m_dnp3)
+	{
+		bool event = info.isEventVariation == true;
+		int flag = static_cast<int>(value.flags.value);
+		bool isBinary = objectType.compare("Binary") == 0 ||
+				objectType.compare("BinaryOutputStatus") == 0;
+
+		// 0x01 means ONLINE for all Objects
+		// STATE is checked for Binary and BinaryOutputStatus objects
+		if (flag == 0x01 ||
```
Comment (MarkRiddoch) [inline dnp3.cpp]: We should maybe create an enum or have a set of defines for the flag values

File: plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+#define DEFER(x) x
+#define TO_STRING_(...) #__VA_ARGS__
+#define QUOTE(...) TO_STRING(__VA_ARGS__)
+
+/**
+ * Default configuration
+ */
+const char *default_config = QUOTE({
+	"plugin" : {
+       		"description" : "Simple DNP3 data change plugin",
```
Comment (MarkRiddoch) [inline plugin.cpp]: minor point - formatting

File: plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+		"type" : "string",
+	       	"default" : "dnp3",
+		"readonly" : "true"
+		},
+	"asset" : {
+       		"description" : "Asset name prefix",
+		"type" : "string",
+	       	"default" : DEFAULT_ASSETNAME_PREFIX,
+		"displayName" : "Asset Name prefix",
```
Comment (MarkRiddoch) [inline plugin.cpp:52]: Display names appear in GUI so need to be short. Descriptions should be more descriptive than display names.

File: plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	       	"order" : "5"
+       		},
+	"outstation_poll_enable" : {
+		"description" : "Enable outstation data poll",
+		"type" : "boolean",
+		"default" : "false",
+		"displayName" : "Poll outstation",
+		"order" : "6"
+		},
+	"outstation_poll_interval" : {
```
Comment (MarkRiddoch) [inline plugin.cpp]: I think it will be consign that we have two poll intervals defined.

File: plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	// Return object
+	return (PLUGIN_HANDLE)dnp3;
+}
+
+/**
+ * Start the Async handling for the plugin
+ */
+void plugin_start(PLUGIN_HANDLE *handle)
```
Comment (MarkRiddoch) [inline plugin.cpp:154]: This should be PLUGIN_HANDLE handle

File: plugin.cpp [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+		throw runtime_error("DNP3 plugin failed to instantiate DNP3 master "
+				    "in 'plugin_start' call");
+		return;
+	}
+}
+
+/**
+ * Register ingest callback
+ */
+void plugin_register_ingest(PLUGIN_HANDLE *handle, INGEST_CB cb, void *data)
```
Comment (MarkRiddoch) [inline plugin.cpp]: Should be PLUGIN_HANDLE, not PLUGIN_HANDLE *

File: plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+				    "in 'plugin_register_ingest' call");
+	}
+	DNP3* dnp3 = (DNP3 *)handle;
+	dnp3->registerIngest(data, cb);
+}
+
+/**
+ * Poll for a plugin reading
+ */
+Reading plugin_poll(PLUGIN_HANDLE *handle)
```
Comment (MarkRiddoch) [inline plugin.cpp]: PLUGIN_HANDLE not PLUGIN_HANDLE *

File: plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+		dnp3->configure(&config);
+		// Start master and connect to outstation
+		dnp3->start();
+	}
+}
+
+/**
+ * Shutdown the DNP3 plugin
+ */
+void plugin_shutdown(PLUGIN_HANDLE *handle)
```
Comment (MarkRiddoch) [inline plugin.cpp]: PLUGIN_HANDLE not PLUGIN_HANDLE *

File: plugin.cpp [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+	throw runtime_error("DNP3 is an async plugin, poll should not be called");
+}
+
+/**
+ * Reconfigure the plugin
+ *
+ */
+void plugin_reconfigure(PLUGIN_HANDLE *handle, string& newConfig)
+{
+	DNP3* dnp3 = (DNP3 *)handle;
```
Comment (MarkRiddoch) [inline plugin.cpp]: DNP3 *dnp = (DNP3 *)*handle

This is the one place it really is a PLUGIN_HANDLE * - at least until we fix that bug

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]
Looks good on the whole

File: README.rst (hunk 2 above)
Comment (pintomax) [reply README.rst]: Yes, oversight

File: dnp3.cpp (hunk 3 above)
Comment (pintomax) [reply dnp3.cpp]: Assuming service name is DNP3_ABC

the label is now: serviceName + _ + "remote_" + outstation id

DNP3_ABC_remote_20

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**